New Feature: AbstractThrowableAssert#message()  #2378

trangntt-016 opened the pull request:
#### Check List:
* Fixes #2378
* Unit tests : YES 
* Javadoc with a code example (on API only) : YES 
* PR meets the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md)

Following the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md) will make it easier for us to review and accept your PR.

scordio commented:
@trangntt-016 as Hacktoberfest will end soon, I want to make sure this PR is counted by the event, so I will label this PR as accepted.

trangntt-016 commented:
Hi @scordio ,

Thanks for the adding. Just a quick question for the tests if you don't mind. I tried to clone assertj-core, then run mvn test and it throws me some exceptions though I didn't touch the codebase. I wonder if it's the reason why my pull request didn't pass Github's checks. Could you please let me know how to fix it? Thanks
https://ibb.co/NNgxhdq

scordio commented on src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:
```diff
@@ -769,4 +769,23 @@ public void doesNotThrowAnyException() {
+   *
+   * // assertion will fail if the provided {@link ThrowableAssert.ThrowingCallable} does not raise an exception, an error is immediately thrown
+   *
+   * @param
+   * @return the created {@link StringAssert}.
```
The build fails because of the unused `@param`. Removing it should fix the checks.

```suggestion
   * @return the created {@link StringAssert}.
```

scordio commented on src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:
```diff
@@ -769,4 +769,23 @@ public void doesNotThrowAnyException() {
+   * @param
+   * @return the created {@link StringAssert}.
+   */
+  public StringAssert message() {
```
Favor abstract type for extendability:

```suggestion
  public AbstractStringAssert<?> message() {
```

trangntt-016 commented:
Hi @scordio 

Thank you for your feedback. I've just updated the code and removed @param. However, there's still one check (receive / store-pitest-feedback (pull_request) that I didn't pass. Would appreciate it if you could gimme some hints on how to what is that check about and how to fix it. Thanks

scordio commented:
We have a general issue with pitest, so you can ignore the failure.

trangntt-016 commented:
> We have a general issue with pitest, so you can ignore the failure.

Thanks @scordio

joel-costigliola commented:
@trangntt-016 could you add soft assertions and assumptions tests using the new `message()` method?
They will likely fail as we must also add `message` to `SoftProxies.METHODS_CHANGING_THE_OBJECT_UNDER_TEST` list.

trangntt-016 commented:
Hi @joel-costigliola ,

Thank you for your feedback. I already added the test using message(). I'm not sure if these are the tests that you meant??
 ```
 @Test
  void should_create_failure_with_no_exception() {
    // GIVEN
    Object actual = null;
    // WHEN
    AbstractStringAssert assertionError = assertThatThrownBy(() -> assertThat(actual).isInstanceOf(Object.class)).message();
    // THEN
    then(assertionError).withFailMessage("Expecting actual not to be null");
  }

  @Test
  void should_pass_with_exception() {
    // GIVEN
    ConcreteAssert assertion = new ConcreteAssert("foo");
    Object actual = "Actual";
    Object expected = "Expected";
    // WHEN
    AssertionFailedError afe = assertion.failureWithActualExpected(actual, expected, "fail");
    // THEN
    then(afe).message().containsAnyOf("fail", "Fiali");
  }

```

joel-costigliola commented:
Integrated thanks @trangntt-016!
